Training/interface/DataLoader_main.h

```diff
@@ -190,10 +190,10 @@ class DataLoader {
         input_histogram .th2d_add(*(input_th2d .get()));
 
         target_histogram.divide(input_histogram);
-        hist_weights[tau_type] = std::make_shared<TH2D>(target_histogram.get_weights_th2d(
+        hist_weights[tau_type] = target_histogram.get_weights_th2d(
```

Comment:
Is this change needed because of the scope?

Reply:
Yes

Reply:
the problem is that [get_weigths_th2d](https://github.com/cms-tau-pog/TauMLTools/blob/ed7b7b385219fa6a239bf972502cb8edf92754d1/Training/interface/histogram2d.h#L176) (as it is in master) created shared pointer inside, but returned a reference:
```c++
TH2D& Histogram_2D::get_weights_th2d(const char* name, const char* title) {
  ...
  histo_ = std::shared_ptr<TH2D>(new TH2D(name, title, nx, xmin_, xmax_, ny, yaxis_.front(), yaxis_.back()));
  return *histo_;
}
```

So the original `histo_` was destructed before the copy constructor is called...